File: apps/worker/src/notification-jobs.ts
import type { NotificationEmailMessage } from "@renewlet/shared/email-template";
import { NOTIFICATION_CHANNELS } from "@renewlet/shared/runtime";
import { notificationJobResultResponseSchema } from "@renewlet/shared/schemas/notifications";
import type { ApiAppSettings } from "@renewlet/shared/schemas/settings";
import type { UpstreamErrorDetails } from "@renewlet/shared/schemas/upstream";
import { NOTIFICATION_JOB_COLUMNS, newId, nowIso, parseJobResult } from "./db";
import type { AppLocale } from "./http";
import type { Env, NotificationJobRow } from "./types";
import type { ScheduleOccurrence } from "./notification-schedule";
import { notificationMessageStatements, splitNotificationJobMessage } from "./notification-message-storage";

// Worker 不能读取 Go 的通知 env；Cloudflare 调度常量固定在这里，并由 shared fixture 与 Go 测试锁住。
export const NOTIFICATION_CRON_WINDOW_MINUTES = 2;
export const NOTIFICATION_MAX_RETRIES = 3;
export const NOTIFICATION_STALE_SENDING_MINUTES = 15;

export type Channel = ApiAppSettings["enabledChannels"][number];

export interface ChannelFailure {
  channel: Channel;
  error: string;
  details?: UpstreamErrorDetails;
}

export interface JobChannels {
  attempted: Channel[];
  succeeded: Channel[];
  failed: ChannelFailure[];
}

export interface SendSummary extends JobChannels {}

const knownChannels = new Set<string>(NOTIFICATION_CHANNELS);

export async function getNotificationJob(env: Env, userId: string, schedule: ScheduleOccurrence): Promise<NotificationJobRow | null> {
  // 这组字段必须与 D1 唯一索引一致，是 cron 幂等和失败重试的同一个业务 key。
  return await env.DB.prepare(`
    SELECT ${NOTIFICATION_JOB_COLUMNS} FROM notification_jobs
    WHERE user_id = ? AND scheduled_local_date = ? AND scheduled_local_time = ? AND time_zone = ?
    LIMIT 1
  `).bind(userId, schedule.scheduledLocalDate, schedule.scheduledLocalTime, schedule.timeZone).first<NotificationJobRow>();
}

export async function createNotificationJob(
  env: Env,
  userId: string,
  schedule: ScheduleOccurrence,
  status: NotificationJobRow["status"],
  attempts: number,
): Promise<{ row: NotificationJobRow | null; created: boolean }> {
  const timestamp = nowIso();
  const id = newId("job");
  // INSERT OR IGNORE 让并发 Cron 只抢到一个发送者；抢不到时回读现有 job 交给状态机判断。
  const result = await env.DB.prepare(`
    INSERT OR IGNORE INTO notification_jobs (
      id, user_id, scheduled_local_date, scheduled_local_time, time_zone, scheduled_instant_utc,
      status, attempts, last_error, result_json, created_at, updated_at
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, NULL, '{}', ?, ?)
  `).bind(id, userId, schedule.scheduledLocalDate, schedule.scheduledLocalTime, schedule.timeZone, schedule.scheduledInstantUtc, status, attempts, timestamp, timestamp).run();
  if ((result.meta.changes ?? 0) === 0) {
    return { row: await getNotificationJob(env, userId, schedule), created: false };
  }
  return {
    row: {
      id,
      user_id: userId,
      scheduled_local_date: schedule.scheduledLocalDate,
      scheduled_local_time: schedule.scheduledLocalTime,
      time_zone: schedule.timeZone,
      scheduled_instant_utc: schedule.scheduledInstantUtc,
      status,
      attempts,
      last_error: null,
      result_json: "{}",
      created_at: timestamp,
      updated_at: timestamp,
    },
    created: true,
  };
}

export async function markNotificationJobSending(env: Env, row: NotificationJobRow, attempts: number): Promise<NotificationJobRow> {
  const timestamp = nowIso();
  // failed/stale sending 接管时先清空 last_error；最终失败摘要只由本次发送结果重新生成。
  await env.DB.prepare(`
    UPDATE notification_jobs SET status = 'sending', attempts = ?, last_error = NULL, updated_at = ?
    WHERE user_id = ? AND id = ?
  `).bind(attempts, timestamp, row.user_id, row.id).run();
  return { ...row, status: "sending", attempts, last_error: null, updated_at: timestamp };
}

export async function finalizeNotificationJob(
  env: Env,
  row: NotificationJobRow | null,
  userId: string,
  schedule: ScheduleOccurrence,
  status: NotificationJobRow["status"],
  attempts: number,
  error: string | null,
  result: unknown,
): Promise<void> {
  let target = row;
  if (!target) {
    // skipped 也需要历史行，否则用户只能看到“没有发送”，看不到本轮 Cron 已检查过。
    const created = await createNotificationJob(env, userId, schedule, status, Math.max(1, attempts));
    target = created.row;
  }
  const timestamp = nowIso();
  const persistedResult = stripNotificationFailureDetails(result);
  if (!target || target.user_id !== userId) throw new Error("Notification job owner mismatch");
  const snapshot = splitNotificationJobMessage(persistedResult);
  // finalize 按调度唯一键更新而不是只按 id，确保 INSERT OR IGNORE 抢占后的同一窗口仍能幂等落最终态。
  const finalize = env.DB.prepare(`
    UPDATE notification_jobs SET status = ?, attempts = ?, last_error = ?, result_json = ?, updated_at = ?
    WHERE user_id = ? AND scheduled_local_date = ? AND scheduled_local_time = ? AND time_zone = ?
  `).bind(status, Math.max(0, attempts), error, snapshot.metadata, timestamp, userId, schedule.scheduledLocalDate, schedule.scheduledLocalTime, schedule.timeZone);
  // D1 batch 统一提交消息与最终态；失败时保留上次完整快照和渠道记录，不能部分替换。
  await env.DB.batch([...notificationMessageStatements(env, target.id, snapshot.parts), finalize]);
}

function stripNotificationFailureDetails(result: unknown): unknown {
  if (!result || typeof result !== "object" || Array.isArray(result)) return result;
  const record = result as Record<string, unknown>;
  const channels = record["channels"];
  if (!channels || typeof channels !== "object" || Array.isArray(channels)) return result;
  const channelRecord = channels as Record<string, unknown>;
  const failed = channelRecord["failed"];
  if (!Array.isArray(failed)) return result;
  return {
    ...record,
    channels: {
      ...channelRecord,
      failed: failed.map((item) => {
        if (!item || typeof item !== "object" || Array.isArray(item)) return item;
        const { details: _details, ...rest } = item as Record<string, unknown>;
        return rest;
      }),
    },
  };
}

export function isNotificationJobTerminal(row: NotificationJobRow | null): boolean {
  return row?.status === "sent" || row?.status === "skipped";
}

export function isSendingJobFresh(row: NotificationJobRow, now: Date, staleSendingMinutes: number): boolean {
  if (row.status !== "sending") return false;
  const updatedAt = Date.parse(row.updated_at);
  if (Number.isNaN(updatedAt)) return false;
  // sending 可能只是上一轮外部 provider 还没返回；未过 stale 窗口不能抢占，避免重复推送。
  return now.getTime() - updatedAt < Math.max(1, staleSendingMinutes) * 60_000;
}

export function readJobChannels(row: NotificationJobRow | null): JobChannels {
  if (!row) return normalizeJobChannels({});
  const result = parseJobResult(row);
  if (!isRecord(result) || result["source"] !== "cron" || !isRecord(result["channels"])) {
    // 手动运行不写 job；坏历史 result 不能参与失败重试，避免把未知 JSON 当作发送依据。
    return normalizeJobChannels({});
  }
  return normalizeJobChannels(result["channels"]);
}

export function channelsToSend(existing: NotificationJobRow | null, previous: JobChannels, enabled: Channel[]): Channel[] {
  if (existing?.status !== "failed") return uniqueValidChannels(enabled);
  // 失败重试只发上次失败且仍启用的渠道，已成功渠道不应因为重试收到重复提醒。
  const enabledSet = new Set(enabled);
  return uniqueValidChannels(previous.failed.map((failure) => failure.channel).filter((channel) => enabledSet.has(channel)));
}

export function mergeChannelResults(previous: JobChannels, summary: SendSummary, enabled: Channel[]): JobChannels {
  // 合并时保留历史成功、替换本轮失败；被禁用渠道的旧失败不再阻塞 job 收敛。
  const sentThisRun = new Set(summary.attempted);
  const succeeded = uniqueValidChannels([...previous.succeeded, ...summary.succeeded]);
  const succeededSet = new Set(succeeded);
  const enabledSet = new Set(enabled);
  const failures = new Map<Channel, string>();
  for (const failure of previous.failed) {
    if (!enabledSet.has(failure.channel)) continue;
    if (sentThisRun.has(failure.channel)) continue;
    if (succeededSet.has(failure.channel)) continue;
    failures.set(failure.channel, failure.error);
  }
  for (const failure of summary.failed) {
    if (succeededSet.has(failure.channel)) continue;
    failures.set(failure.channel, failure.error);
  }
  return {
    attempted: uniqueValidChannels([...previous.attempted, ...summary.attempted]),
    succeeded,
    failed: [...failures.entries()]
      .map(([channel, error]) => ({ channel, error }))
      .sort((a, b) => a.channel.localeCompare(b.channel)),
  };
}

export function normalizeJobChannels(value: unknown): JobChannels {
  const record = isRecord(value) ? value : {};
  const failedByChannel = new Map<Channel, string>();
  const rawFailed = Array.isArray(record["failed"]) ? record["failed"] : [];
  for (const item of rawFailed) {
    if (!isRecord(item) || typeof item["channel"] !== "string") continue;
    const channel = parseChannel(item["channel"]);
    if (!channel) continue;
    failedByChannel.set(channel, typeof item["error"] === "string" ? item["error"] : "");
  }
  return {
    attempted: uniqueValidChannels(Array.isArray(record["attempted"]) ? record["attempted"] : []),
    succeeded: uniqueValidChannels(Array.isArray(record["succeeded"]) ? record["succeeded"] : []),
    failed: [...failedByChannel.entries()]
      .map(([channel, error]) => ({ channel, error }))
      .sort((a, b) => a.channel.localeCompare(b.channel)),
  };
}

export function lastErrorFromChannels(channels: JobChannels): string | null {
  if (channels.failed.length === 0) return null;
  return channels.failed.map((failure) => `${failure.channel}: ${failure.error}`).join(" | ");
}

export function publicScheduleOccurrence(schedule: ScheduleOccurrence): ScheduleOccurrence {
  // due/reason 只属于 Cron 内存决策；notification_jobs.result_json 必须落 shared 公开 occurrence 契约。
  return {
    scheduledLocalDate: schedule.scheduledLocalDate,
    scheduledLocalTime: schedule.scheduledLocalTime,
    timeZone: schedule.timeZone,
    scheduledInstantUtc: schedule.scheduledInstantUtc,
  };
}

export function normalizeNotificationJobResultForHistory(result: unknown): unknown {
  // history 读路径只验收当前 shared cron result；旧/坏 result 不再运行时重塑，统一以 `{}` 出站。
  const parsed = notificationJobResultResponseSchema.safeParse(result);
  return parsed.success ? parsed.data : {};
}

export function createCronJobResult(input: {
  reason: string | null;
  force: boolean;
  windowMinutes: number;
  triggeredAtUtc: string;
  schedule: ScheduleOccurrence;
  settings: ApiAppSettings;
  locale: AppLocale;
  message: NotificationEmailMessage;
  channels: JobChannels;
}): unknown {
  // 历史 result 只保存可解释的 cron 快照，不写 provider token、完整外部响应或 manual source。
  return {
    source: "cron",
    reason: input.reason,
    force: input.force,
    windowMinutes: input.windowMinutes,
    triggeredAtUtc: input.triggeredAtUtc,
    schedule: publicScheduleOccurrence(input.schedule),
    settings: {
      timezone: input.settings.timezone,
      locale: input.locale,
      notificationTimeLocal: input.settings.notificationTimeLocal,
      enabledChannels: input.settings.enabledChannels,
      showExpired: input.settings.showExpired,
    },
    message: input.message,
    channels: normalizeJobChannels(input.channels),
  };
}

function uniqueValidChannels(values: unknown[]): Channel[] {
  const out: Channel[] = [];
  const seen = new Set<string>();
  for (const value of values) {
    const channel = typeof value === "string" ? parseChannel(value) : null;
    if (!channel || seen.has(channel)) continue;
    seen.add(channel);
    out.push(channel);
  }
  return out;
}

function parseChannel(value: string): Channel | null {
  return knownChannels.has(value) ? value as Channel : null;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: apps/docker-server/cmd/renewlet/notification_jobs.go
package main

// notification_jobs 拥有调度幂等和渠道重试元数据；完整消息只存私有快照表，由历史查询组装公开 DTO。
//
// 注意： 唯一索引冲突被视为并发执行已抢占；修改这里的错误处理会直接影响重复发送保护。
import (
	"bytes"
	"encoding/json"
	"errors"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
)

func getNotificationJob(app core.App, userID, localDate, localTime, timezone string) (*core.Record, error) {
	// 这组字段与 notification_jobs 唯一索引一致，是调度幂等的查找边界。
	return app.FindFirstRecordByFilter(
		"notification_jobs",
		"user = {:user} && scheduledLocalDate = {:date} && scheduledLocalTime = {:time} && timeZone = {:tz}",
		dbx.Params{"user": userID, "date": localDate, "time": localTime, "tz": timezone},
	)
}

// createNotificationJob 创建通知任务并返回是否由本次调用创建。
// 注意： 唯一索引冲突被视为“其他并发执行已创建”，调用方应跳过本用户。
func createNotificationJob(app core.App, userID string, schedule localScheduleDecision, status string, attempts int) (*core.Record, bool, error) {
	if existing, err := getNotificationJob(app, userID, schedule.ScheduledLocalDate, schedule.ScheduledLocalTime, schedule.TimeZone); err == nil && existing != nil {
		return existing, false, nil
	}
	collection, err := app.FindCollectionByNameOrId("notification_jobs")
	if err != nil {
		return nil, false, err
	}
	record := core.NewRecord(collection)
	record.Set("user", userID)
	record.Set("scheduledLocalDate", schedule.ScheduledLocalDate)
	record.Set("scheduledLocalTime", schedule.ScheduledLocalTime)
	record.Set("timeZone", schedule.TimeZone)
	record.Set("scheduledInstantUtc", schedule.ScheduledInstantUTC)
	record.Set("status", status)
	record.Set("attempts", attempts)
	record.Set("lastError", "")
	record.Set("result", emptyJSONPayload{})
	if err := app.Save(record); err != nil {
		if strings.Contains(strings.ToLower(err.Error()), "unique") {
			return nil, false, nil
		}
		return nil, false, err
	}
	return record, true, nil
}

func markNotificationJobSending(app core.App, record *core.Record, attempts int) error {
	record.Set("status", notificationStatusSending)
	record.Set("attempts", attempts)
	record.Set("lastError", "")
	return app.Save(record)
}

// finalizeNotificationJob 写入任务最终状态和严格 result。
func finalizeNotificationJob(app core.App, record *core.Record, userID string, schedule localScheduleDecision, status string, lastError string, result notificationJobResult) error {
	payload, err := json.Marshal(result.Message)
	if err != nil {
		return err
	}
	parts := notificationMessageParts(payload)
	metadata := notificationJobStoredResult{notificationJobMetadata: result.notificationJobMetadata, MessageChunkCount: len(parts)}
	// 渠道完成状态和完整消息快照是一次提交；任一写入失败都不能留下新状态配半份正文。
	return app.RunInTransaction(func(txApp core.App) error {
		var updated *core.Record
		var err error
		if record == nil {
			updated, _, err = createNotificationJob(txApp, userID, schedule, status, 1)
		} else {
			updated, err = txApp.FindRecordById("notification_jobs", record.Id)
		}
		if err != nil {
			return err
		}
		if updated == nil {
			return errors.New("notification job was not created")
		}
		if updated.GetString("user") != userID {
			return errors.New("notification job owner mismatch")
		}
		if err := writeNotificationJobMessage(txApp, updated.Id, parts); err != nil {
			return err
		}
		updated.Set("status", status)
		updated.Set("lastError", lastError)
		updated.Set("result", metadata)
		return txApp.Save(updated)
	})
}

// createJobResult 构造历史面板可解析的 cron result。
// 注意： 新增字段时必须同步前端 cronJobResultResponseSchema。
func createJobResult(reason string, schedule localScheduleOccurrence, settings appSettings, locale appLocale, due notificationMessage, options notificationCronOptions, channels jobChannels) notificationJobResult {
	var reasonValue *string
	if reason != "" {
		reasonValue = &reason
	}
	return normalizeNotificationJobResult(notificationJobResult{
		notificationJobMetadata: notificationJobMetadata{
			Source:         "cron",
			Reason:         reasonValue,
			Force:          options.Force,
			WindowMinutes:  options.WindowMinutes,
			TriggeredAtUTC: options.Now.UTC().Format(time.RFC3339),
			Schedule:       schedule,
			Settings: notificationJobResultSettings{
				Timezone:              settings.Timezone,
				Locale:                string(locale),
				NotificationTimeLocal: settings.NotificationTimeLocal,
				EnabledChannels:       settings.EnabledChannels,
				ShowExpired:           settings.ShowExpired,
			},
			Channels: channels,
		},
		Message: notificationJobResultMessage{
			Title:      due.Title,
			Content:    due.Content,
			Timestamp:  due.Timestamp,
			HasPayload: due.HasPayload,
			Items:      due.Items,
		},
	})
}

// readJobChannels 从历史 result 中读取渠道状态，用于失败重试合并。
func readJobChannels(record *core.Record) jobChannels {
	var result notificationJobStoredResult
	if err := decodeJSONRecordField(record, "result", &result); err != nil {
		return normalizeJobChannels(jobChannels{})
	}
	if result.Source != "cron" {
		return normalizeJobChannels(jobChannels{})
	}
	return normalizeJobChannels(result.Channels)
}

// normalizeJobChannels 保证渠道集合输出稳定为 JSON array，而不是 nil slice 对应的 null。
func normalizeJobChannels(channels jobChannels) jobChannels {
	failedByChannel := map[string]string{}
	for _, failure := range channels.Failed {
		channel := strings.TrimSpace(failure.Channel)
		if _, ok := knownChannels[channel]; !ok {
			continue
		}
		// 同一渠道只保留最后一次错误，防止历史 result 因多次重试无限膨胀。
		failedByChannel[channel] = failure.Error
	}
	failed := make([]channelFailure, 0, len(failedByChannel))
	for channel, errText := range failedByChannel {
		// notification_jobs 是 cron 审计数据，只保存短摘要；上游 raw response 只能留在手动 API 的当前响应里。
		failed = append(failed, channelFailure{Channel: channel, Error: errText})
	}
	sort.Slice(failed, func(i, j int) bool { return failed[i].Channel < failed[j].Channel })
	return jobChannels{
		Attempted: uniqueValidChannels(channels.Attempted),
		Succeeded: uniqueValidChannels(channels.Succeeded),
		Failed:    failed,
	}
}

// normalizeNotificationJobResult 是通知历史 result 的唯一输出契约收敛点。
func normalizeNotificationJobResult(result notificationJobResult) notificationJobResult {
	result.Settings.EnabledChannels = uniqueValidChannels(result.Settings.EnabledChannels)
	if result.Message.Items == nil {
		result.Message.Items = []notificationContentItem{}
	}
	result.Channels = normalizeJobChannels(result.Channels)
	return result
}

// channelsToSend 返回本轮实际需要发送的渠道。
// 对失败重试只发送仍启用且上次失败的渠道，防止成功渠道重复通知。
func channelsToSend(existing *core.Record, previous jobChannels, enabled []string) []string {
	if existing != nil && existing.GetString("status") == notificationStatusFailed {
		enabledSet := map[string]struct{}{}
		for _, channel := range enabled {
			enabledSet[channel] = struct{}{}
		}
		out := []string{}
		for _, failure := range previous.Failed {
			if _, ok := enabledSet[failure.Channel]; ok {
				out = append(out, failure.Channel)
			}
		}
		return uniqueValidChannels(out)
	}
	return enabled
}

// mergeChannelResults 合并历史渠道结果和本轮发送结果。
// 已成功渠道优先，已禁用渠道的失败记录会被丢弃，避免历史错误阻塞新设置。
func mergeChannelResults(previous jobChannels, summary sendSummary, enabled []string) jobChannels {
	sentThisRun := map[string]struct{}{}
	for _, channel := range summary.Attempted {
		sentThisRun[channel] = struct{}{}
	}
	succeeded := uniqueValidChannels(append(previous.Succeeded, summary.Succeeded...))
	succeededSet := map[string]struct{}{}
	for _, channel := range succeeded {
		succeededSet[channel] = struct{}{}
	}
	enabledSet := map[string]struct{}{}
	for _, channel := range enabled {
		enabledSet[channel] = struct{}{}
	}
	failures := map[string]string{}
	for _, failure := range previous.Failed {
		if _, ok := enabledSet[failure.Channel]; !ok {
			continue
		}
		if _, ok := sentThisRun[failure.Channel]; ok {
			continue
		}
		if _, ok := succeededSet[failure.Channel]; ok {
			continue
		}
		failures[failure.Channel] = failure.Error
	}
	for _, failure := range summary.Failed {
		if _, ok := succeededSet[failure.Channel]; ok {
			continue
		}
		failures[failure.Channel] = failure.Error
	}
	failed := make([]channelFailure, 0, len(failures))
	for channel, errText := range failures {
		// merge 结果会进入 cron history，必须丢弃手动/本轮发送捕获的 raw details。
		failed = append(failed, channelFailure{Channel: channel, Error: errText})
	}
	sort.Slice(failed, func(i, j int) bool { return failed[i].Channel < failed[j].Channel })
	return jobChannels{
		Attempted: uniqueValidChannels(append(previous.Attempted, summary.Attempted...)),
		Succeeded: succeeded,
		Failed:    failed,
	}
}

func decodeJSONRecordField(record *core.Record, field string, target interface{}) error {
	value := record.Get(field)
	switch v := value.(type) {
	case string:
		if strings.TrimSpace(v) == "" {
			return nil
		}
		return json.Unmarshal([]byte(v), target)
	case []byte:
		if len(bytes.TrimSpace(v)) == 0 {
			return nil
		}
		return json.Unmarshal(v, target)
	default:
		// PocketBase JSON 字段在 hooks/routes/tests 中可能是 map、struct 或 RawMessage；统一 marshal 后再解码到目标类型。
		data, err := json.Marshal(v)
		if err != nil {
			return err
		}
		return json.Unmarshal(data, target)
	}
}

// summarizeCronResult 汇总单次 cron 执行结果。
func summarizeCronResult(options notificationCronOptions, results []notificationCronUserResult) notificationCronResult {
	out := notificationCronResult{
		NowUTC:    options.Now.UTC().Format(time.RFC3339),
		Force:     options.Force,
		DryRun:    options.DryRun,
		Processed: len(results),
		Results:   results,
	}
	for _, result := range results {
		switch result.Action {
		case "sent":
			out.Sent++
		case "failed":
			out.Failed++
		case "skipped":
			out.Skipped++
		}
	}
	return out
}

// normalizeNotificationHistoryResult 只验收重建后的公开契约，不保留旧形状的运行时转换。
func normalizeNotificationHistoryResult(data []byte) json.RawMessage {
	trimmed := bytes.TrimSpace(data)
	if bytes.Equal(trimmed, []byte("{}")) {
		return json.RawMessage("{}")
	}
	var result notificationJobResult
	if err := decodeStrictJSONBytesInto(trimmed, &result, defaultAppLocale, false); err != nil {
		return json.RawMessage("{}")
	}
	// history 读路径只验收当前 cron result wire shape；旧/坏历史不再运行时重塑。
	if !notificationJobResultIsCurrentContract(result) {
		return json.RawMessage("{}")
	}
	return json.RawMessage(trimmed)
}

func notificationJobResultIsCurrentContract(result notificationJobResult) bool {
	if result.Source != "cron" || result.WindowMinutes < 0 {
		return false
	}
	if !isValidDateOnly(result.Schedule.ScheduledLocalDate) ||
		!isValidLocalTime(result.Schedule.ScheduledLocalTime) ||
		strings.TrimSpace(result.Schedule.TimeZone) == "" ||
		strings.TrimSpace(result.Schedule.ScheduledInstantUTC) == "" {
		return false
	}
	if !isValidLocalTime(result.Settings.NotificationTimeLocal) ||
		result.Settings.EnabledChannels == nil ||
		result.Message.Items == nil ||
		result.Channels.Attempted == nil ||
		result.Channels.Succeeded == nil ||
		result.Channels.Failed == nil {
		return false
	}
	if !allKnownChannels(result.Settings.EnabledChannels) || !jobChannelsUseCurrentContract(result.Channels) {
		return false
	}
	for _, item := range result.Message.Items {
		if !notificationContentItemUsesCurrentContract(item) {
			return false
		}
	}
	return true
}

func jobChannelsUseCurrentContract(channels jobChannels) bool {
	if !allKnownChannels(channels.Attempted) || !allKnownChannels(channels.Succeeded) {
		return false
	}
	for _, failure := range channels.Failed {
		if _, ok := knownChannels[failure.Channel]; !ok {
			return false
		}
	}
	return true
}

func notificationContentItemUsesCurrentContract(item notificationContentItem) bool {
	switch item.Type {
	case "renewal", "trial", "expired", "expiry":
	default:
		return false
	}
	if !isValidSubscriptionStatus(item.Status) || !isValidDateOnly(item.TargetDate) || item.ReminderDays < 0 {
		return false
	}
	if item.RepeatReminder != nil &&
		(!isValidRepeatReminderInterval(item.RepeatReminder.Interval) || !isValidRepeatReminderWindow(item.RepeatReminder.Window)) {
		return false
	}
	return true
}

func allKnownChannels(channels []string) bool {
	for _, channel := range channels {
		if _, ok := knownChannels[channel]; !ok {
			return false
		}
	}
	return true
}

func nullableString(value string) *string {
	if value == "" {
		return nil
	}
	return &value
}

func parseInt(value string, fallback int) int {
	if strings.TrimSpace(value) == "" {
		return fallback
	}
	parsed, err := strconv.Atoi(value)
	if err != nil {
		return fallback
	}
	return parsed
}

func clampInt(value, min, max int) int {
	if value < min {
		return min
	}
	if value > max {
		return max
	}
	return value
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}
